topologies/multi.go

```diff
@@ -109,45 +109,56 @@ func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes ma
 
 	rrIndex := -1
 	rrIndexPerZone := map[string]int{}
+	rrPerZone := map[string][]*corev1.Node{}
+
+	// Creat per zone RR lists for MZR selection
+	if t.Config.ZoneLabel != "" {
+		for i, rr := range routeReflectors {
+			rrZone := rr.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("RR:%s's zone: %s", rr.Name, rrZone)
+			rrPerZone[rrZone] = append(rrPerZone[rrZone], &routeReflectors[i])
+		}
+	}
 
 	for _, n := range nodeList {
 		if t.IsRouteReflector(string(n.GetUID()), n.GetLabels()) {
 			continue
 		}
 
-		routeReflectorsForNode := []corev1.Node{}
+		routeReflectorsForNode := []*corev1.Node{}
 
+		// Run MZR selection first
 		if t.Config.ZoneLabel != "" {
-			log.Debugf("Node's zone: %s", n.GetLabels()[t.Config.ZoneLabel])
-
 			nodeZone := n.GetLabels()[t.Config.ZoneLabel]
-			rrsSameZone := []corev1.Node{}
-
-			for i, rr := range routeReflectors {
-				rrZone := rr.GetLabels()[t.Config.ZoneLabel]
-				log.Debugf("RR:%s's zone: %s", rr.Name, rrZone)
-				if nodeZone == rrZone {
-					if _, ok := rrIndexPerZone[nodeZone]; !ok {
-						rrIndexPerZone[nodeZone] = -1
-					}
-
-					log.Debugf("Adding RR:%s to Same Zone RRs", rr.Name)
-					rrsSameZone = append(rrsSameZone, routeReflectors[i])
-				}
-			}
+			log.Debugf("Node's zone: %s", nodeZone)
 
-			if len(rrsSameZone) > 0 {
+			// Select the 1st RR from the same zone if there're any
+			if len(rrPerZone[nodeZone]) > 0 {
 				rrIndexPerZone[nodeZone]++
-				if rrIndexPerZone[nodeZone] == len(rrsSameZone) {
+				if rrIndexPerZone[nodeZone] == len(rrPerZone[nodeZone]) {
 					rrIndexPerZone[nodeZone] = 0
 				}
 
-				rr := rrsSameZone[rrIndexPerZone[nodeZone]]
+				rr := rrPerZone[nodeZone][rrIndexPerZone[nodeZone]]
 				log.Debugf("Adding %s to RRs for Node:%s", rr.Name, n.Name)
 				routeReflectorsForNode = append(routeReflectorsForNode, rr)
 			}
+
+			// Select the 2nd RR from a different zone
+			for zone := range rrPerZone {
+				if zone != nodeZone {
```

Comment:
Am i right that maximum route reflectors handling is missing here and it will select from all available other zones?

Reply:
The break in line 156 should stop the for loop after it found the first RR from another zone and added it as 2nd RR to the Node. And then the next loop in line 162 should ensure that enough RRs are selected in zone-aware and in zone-unaware clusters as well.

Reply:
Eliminated duplicate code in MZR 1st&2nd RR selection, but it's still similar to what it does at https://github.com/mhmxs/calico-route-reflector-operator/pull/9/files#diff-2647a31e99ca8b01345ef411657d0452R155 for sequential selection...